iBeacon support for ble_presence

ignisf opened the pull request:
# What does this implement/fix? 

This PR adds a convenient way to track the presence of iBeacon devices.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes https://github.com/esphome/issues/issues/1912, fixes https://github.com/esphome/issues/issues/1913, fixes https://github.com/esphome/esphome/pull/937, closes esphome/feature-requests#806

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs/pull/1069
  
# Test Environment

- [x] ESP32
- [ ] ~~ESP8266~~ (N/A)
- [ ] Windows
- [ ] Mac OS
- [x] Linux

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml
esp32_ble_tracker:

binary_sensor:
  - platform: ble_presence
    ibeacon_uuid: 77a6438d-ea95-4522-b46c-cb2b4412076f
    ibeacon_major: 100
    ibeacon_minor: 1
    name: "Jane's Phone"

```

# Explain your changes

* The Home Assistant Android application offers the following useful functionality: [Screenshot](https://user-images.githubusercontent.com/1153188/111885321-c027c700-89cf-11eb-8fec-3833bd48709f.png)

* When activated, it starts emitting iBeacon announcements from a randomly changing MAC address and with no service UUIDs.

* ESPHome currently offers no convenient way of detecting the presence of iBeacons (see the related issues). This makes it hard for users to utilize the Home Assistant application for BLE tracking of their devices.

This PR introduces support for iBeacon presence detection in the `ble_presence` component.

After the ground work here is done, it should be relatively simple to mirror the changes to `ble_rssi` as well.

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

ignisf committed:
Introduce logging of received iBeacon data

ignisf committed:
Rename esp32_ble_tracer.as_hex_array to as_reversed_hex_array

The mentioned function actually returns a reversed hex array. The change makes
sure this fact obvious from its name.

ignisf committed:
Introduce iBeacon presence tracking in ble_presence

ignisf commented on esphome/components/ble_presence/binary_sensor.py:
```diff
@@ -19,11 +26,14 @@
             cv.Optional(CONF_SERVICE_UUID): esp32_ble_tracker.bt_uuid,
+            cv.Optional(CONF_IBEACON_MAJOR): cv.uint16_t,
+            cv.Optional(CONF_IBEACON_MINOR): cv.uint16_t,
+            cv.Optional(CONF_IBEACON_UUID): esp32_ble_tracker.ibeacon_uuid,
```
Could I have a suggestion how to enforce `ibeacon_uuid` is set in case any of `ibeacon_major` and `ibeacon_minor` are set?

ignisf replied:
Of course this is assuming that all those identifiers have a hierarchy... I could change it so either of iBeacon UUID/Major/Minor can be specified to perform iBeacon presence detection to keep it simple.

ignisf replied:
I added a `validate` function similar to:
https://github.com/esphome/esphome/blob/808ce6eecba4b086670f87ceb5de1a710a7f4d67/esphome/components/hm3301/sensor.py#L23-L28

rdehuyss commented:
Would love to have this merged as it currently requires extra work in ESPHome. If I just specify the service_uuid for the binary sensor ble_presence, this does not work.

This is how it should be:
```
    - platform: ble_rssi
      service_uuid: '3e5c314b-17d9-4cad-b0db-052cf2de34ee'
      name: Ronald GSM BLE $devicename_friendly
```

This is how I now do it:
```
esp32_ble_tracker:
  on_ble_advertise:
    - then:
        - lambda: |-
            if(x.get_ibeacon().has_value()) {
              auto ibeacon = x.get_ibeacon().value();
              if(strcmp(ibeacon.get_uuid().to_string().c_str(), "C3:D1:26:E2:7E:B1:43:EF:AE:59:82:48:D5:33:17:1C") == 0) {
                id(my_rssi).publish_state(x.get_rssi());
              }
            }

sensor:
  - platform: template
    id: my_rssi
    name: "$Ronald GSM BLE RSSI $devicename_friendly"
    icon: "mdi:mdi-signal-distance-variant"
    update_interval: never
    unit_of_measurement: "dBm"
    filters:
      - sliding_window_moving_average:
          window_size: 5
          send_every: 1
```

ignisf commented:
I've enhanced the config validation and have rebased on the latest `dev`.

federicoodorizzi commented:
When will this feature be available? Thanks!

marcinbauer85 commented:
Any ETA on when this will be live? 🙏

flyize commented:
@rdehuyss Until this gets merged, any way you can explain your fix a bit? Your UUID in the first code block doesn't appear to be represented in your second code block.

Am I just being a bit too literal?

jesserockz commented on esphome/components/esp32_ble_tracker/__init__.py:
```diff
@@ -104,11 +104,40 @@ def bt_uuid(value):
+ibeacon_uuid_format = "XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX"
+
+
+def ibeacon_uuid(value):
+    in_value = cv.string_strict(value)
+    value = in_value.upper()
+
+    if len(value) == len(ibeacon_uuid_format):
+        pattern = re.compile(
+            "^[A-F|0-9]{8,}-[A-F|0-9]{4,}-[A-F|0-9]{4,}-[A-F|0-9]{4,}-[A-F|0-9]{12,}$"
+        )
+        if not pattern.match(value):
+            raise cv.Invalid(
+                f"Invalid hexadecimal value for iBeacon UUID format: '{in_value}'"
+            )
+        return value
+    raise cv.Invalid("iBeacon UUID must be in '{}' format".format(ibeacon_uuid_format))
```
Could you not just use `cv.uuid` instead of writing this?

jesserockz commented on esphome/components/ble_presence/binary_sensor.py:
```diff
@@ -13,17 +20,30 @@
             cv.Optional(CONF_SERVICE_UUID): esp32_ble_tracker.bt_uuid,
+            cv.Optional(CONF_IBEACON_MAJOR): cv.uint16_t,
+            cv.Optional(CONF_IBEACON_MINOR): cv.uint16_t,
+            cv.Optional(CONF_IBEACON_UUID): esp32_ble_tracker.ibeacon_uuid,
```
```suggestion
            cv.Optional(CONF_IBEACON_UUID): cv.uuid,
```

jesserockz commented on esphome/components/ble_presence/binary_sensor.py:
```diff
@@ -13,17 +20,30 @@
 )
 
+
+def validate(config):
```
Please rename this function to be specific.

This is mainly because there is an upcoming change that specifically looks for `validate` functions in components

federicoodorizzi commented:
Has this been implemented?

alexd321 commented on esphome/components/ble_presence/ble_presence_device.h:
```diff
@@ -14,52 +14,99 @@ class BLEPresenceDevice : public binary_sensor::BinarySensorInitiallyOff,
+    this->check_ibeacon_minor_ = true;
+    this->ibeacon_minor_ = minor;
+  }
   void on_scan_end() override {
```
I'm fairly new to this, so sorry I am wrong. I've been testing this PR, I think this would mean state returns to false at the end of _scan_parameters: duration_, correct?
would the be any benefit to having a custom configurable "off time"? or making the duration clear in the docs?

ignisf replied:
> this would mean state returns to false at the end of scan_parameters: duration, correct?

This is my understanding as well.

> would the be any benefit to having a custom configurable "off time"

Well... The off time is already configurable by setting `scan_parameters.duration`?

> or making the duration clear in the docs?

I support this, we need confirmation from someone with more experience with `esp32_ble_tracker` though.

ignisf commented:
Hello @jesserockz and thank you for the review. I did not expect this PR to be so popular. I've addressed your feedback and am currently waiting for CI and testing the changes on a device.

maxiangelo commented:
Great work, this is really needed to get rid of unneeded apps beside the companion app!

jesserockz commented:
Sorry for the delay :rocket:
